xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml

```diff
@@ -976,28 +976,239 @@ XWiki.Annotation = Class.create({
+  */
+  getTextNodeAtPlainTextOffset: function(parentNode, plainTextOffset) {
+    var childNodes = parentNode.childNodes;
+    var child;
+    var parentNodePlainTextLength = 0;
+    for (let i = 0; i &lt; childNodes.length; i++) {
+      child = childNodes[i];
+      if (child &amp;&amp; child.nodeType == 3) {
+        var previousSiblingsLength = parentNodePlainTextLength;
+        // The spaces are ignored since they were removed as well on the server when the offset was computed.
+        parentNodePlainTextLength += child.textContent.replace(/\s/g, '').length;
+        if (plainTextOffset &lt; parentNodePlainTextLength) {
+          // Because plainTextOffset doesn't consider spaces, the real offset of the node needs to be recomputed so that
+          // they are included.
+          return {'node': child, 'offset': this.getNodeSpecificOffset(child, plainTextOffset - previousSiblingsLength)};
+        }
+      }
+      if (child.childNodes.length &gt; 0) {
+        var maybeFoundNode = this.getTextNodeAtPlainTextOffset(child, plainTextOffset - parentNodePlainTextLength);
+        if (maybeFoundNode &amp;&amp; maybeFoundNode.node) {
+          return maybeFoundNode;
+        } else {
+          parentNodePlainTextLength += maybeFoundNode.childNodeLength;
+        }
+      }
+    }
+    return {'childNodeLength': parentNodePlainTextLength};
+  },
+
+  /**
+   * Get a offset knowing the offset relative to the plain text of the element.
+   *
+   * @param node a DOM node
+   * @param offset the offset relative to the plain text
+  */
+  getNodeSpecificOffset: function(node, offset) {
+    var nonSpaceCharsLength = 0;
+    var chars = Array.from(node.textContent);
+    for (let i = 0; i &lt; chars.length; i++) {
+      if (/\s/.test(chars[i])) {
+        continue;
+      } else {
+        if (nonSpaceCharsLength == offset) {
+          return i;
+        }
+        nonSpaceCharsLength += 1;
+      }
+    }
+  },
+
+  /**
+   * Filter only the text nodes inside a DOM Range.
+   *
+   * @param range a DOM Range
+  */
+  getTextNodesInRange: function(range) {
+    var rangeIterator = document.createNodeIterator(
+      range.commonAncestorContainer,
+      NodeFilter.SHOW_TEXT,
+      {
+        acceptNode: function (node) {
+          // Only accept nodes that have content other than whitespace.
+          if (/\S/.test(node.data)) {
+            return NodeFilter.FILTER_ACCEPT
+          }
+        }
+      }
+    );
+    var nodes = [];
+    // Get the Range start and end as text nodes.
+    var startContainer = range.startContainer;
+    if (startContainer.childNodes.length &gt; 0) {
+      startContainer = startContainer.childNodes[range.startOffset];
+    }
+    var endContainer = range.endContainer;
+    if (endContainer.childNodes.length &gt; 0) {
+      // The node at the endOffset position is the boundary point of the Range.
+      endContainer = endContainer.childNodes[range.endOffset - 1];
+    }
```

Comment:
The way you compute the start and end container is error-prone. I'd rather use https://developer.mozilla.org/en-US/docs/Web/API/Range/compareBoundaryPoints inside the range iterator loop to check if the current node is inside the range.